File: src/main/java/com/morph/engine/graphics/Texture.java
package com.morph.engine.graphics;

import com.morph.engine.util.IOUtils;
import org.lwjgl.BufferUtils;
import org.lwjgl.stb.STBTTBakedChar;
import org.lwjgl.stb.STBTruetype;

import javax.imageio.ImageIO;
import java.awt.image.BufferedImage;
import java.io.IOException;
import java.nio.ByteBuffer;
import java.util.HashMap;

import static org.lwjgl.opengl.GL11.*;
import static org.lwjgl.opengl.GL13.GL_TEXTURE0;
import static org.lwjgl.opengl.GL13.glActiveTexture;

public class Texture {
	private TextureResource resource;
	private String filename;

	private static HashMap<String, TextureResource> loadedTextures = new HashMap<>();
	
	public Texture(String filename) {
		this.filename = filename != null ? filename : "textures/solid.png";
		TextureResource oldResource = loadedTextures.get(this.filename);
		if (oldResource != null) {
			this.resource = oldResource;
			resource.addReference();
		} else {
			resource = loadTexture(this.filename);
			loadedTextures.put(this.filename, resource);
		}
	}
	
	public Texture(String font, int size, int bitmapWidth, int bitmapHeight) {
		this.filename = font != null ? font : "C:/Windows/Fonts/arial.ttf";
		TextureResource oldResource = loadedTextures.get(this.filename + ":" + size);
		if (oldResource != null) {
			this.resource = oldResource;
			resource.addReference();
		} else {
			resource = loadFont(this.filename, size, bitmapWidth, bitmapHeight);
			loadedTextures.put(this.filename, resource);
		}
	}

	public Texture(String id, int width, int height, ByteBuffer pixels) {
		this.filename = id;
		TextureResource oldResource = loadedTextures.get(this.filename);
		if (oldResource != null) {
			this.resource = oldResource;
			resource.addReference();
		} else {
			resource = loadTextureFromByteBuffer(width, height, pixels);
			loadedTextures.put(this.filename, resource);
		}
	}

	private TextureResource loadTextureFromByteBuffer(int width, int height, ByteBuffer buffer) {
		int ID = glGenTextures();
		buffer.flip();

		glBindTexture(GL_TEXTURE_2D, ID);

		glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_REPEAT);
		glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_REPEAT);

		glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
		glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);

		glTexImage2D(GL_TEXTURE_2D, 0, GL_ALPHA, width, height, 0, GL_ALPHA, GL_UNSIGNED_BYTE, buffer);

		return new TextureResource(ID);
	}
	
	private TextureResource loadTexture(String filename) {
		try {
			BufferedImage image = ImageIO.read(Texture.class.getClassLoader().getResourceAsStream(filename));
			int[] pixels = image.getRGB(0, 0, image.getWidth(), image.getHeight(), null, 0, image.getWidth());
			
			ByteBuffer buffer = BufferUtils.createByteBuffer(image.getWidth() * image.getHeight() * 4);
			
			boolean hasAlpha = image.getColorModel().hasAlpha();
			
			for (int y = 0; y < image.getHeight(); y++) {
				for (int x = 0; x < image.getWidth(); x++) {
					int color = pixels[x + y * image.getWidth()];
					buffer.put((byte)((color >> 16) & 0xff));
					buffer.put((byte)((color >> 8) & 0xff));
					buffer.put((byte)(color & 0xff));
					
					if (hasAlpha)
						buffer.put((byte)((color >> 24) & 0xff));
					else
						buffer.put((byte)(0xff));
				}
			}
			
			buffer.flip();
			
			int ID = glGenTextures();
			
			glBindTexture(GL_TEXTURE_2D, ID);
			
			glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_REPEAT);
			glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_REPEAT);
			
			glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
			glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
			
			glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA8, image.getWidth(), image.getHeight(), 0, GL_RGBA, GL_UNSIGNED_BYTE, buffer);
			
			return new TextureResource(ID);
		} catch (IOException e) {
			e.printStackTrace();
		}
		
		return null;
	}
	
	private TextureResource loadFont(String font, int size, int bitmapWidth, int bitmapHeight) {
		int texture = glGenTextures();
		STBTTBakedChar.Buffer chars = STBTTBakedChar.malloc(96);
		
		try {
			ByteBuffer fontBuffer = IOUtils.INSTANCE.getFileAsByteBuffer(font, 160 * 1024);
			ByteBuffer bitmap = BufferUtils.createByteBuffer(bitmapWidth * bitmapHeight);
			
			STBTruetype.stbtt_BakeFontBitmap(fontBuffer, size, bitmap, bitmapWidth, bitmapHeight, 32, chars);
			
			glBindTexture(GL_TEXTURE_2D, texture);
			glTexImage2D(GL_TEXTURE_2D, 0, GL_ALPHA, bitmapWidth, bitmapHeight, 0, GL_ALPHA, GL_UNSIGNED_BYTE, bitmap);
			glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
			glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
		} catch (IOException e) {
			throw new RuntimeException(e);
		}
		
		return new TextureResource(texture);
	}

	public void destroy() {
		resource.removeReference();
	}
	
	public void bind() {
		glActiveTexture(GL_TEXTURE0);
		glBindTexture(GL_TEXTURE_2D, resource.getId());
	}

	public void bind(int i) {
		glActiveTexture(GL_TEXTURE0 + i);
		glBindTexture(GL_TEXTURE_2D, resource.getId());
	}
	
	public void unbind() {
		glBindTexture(GL_TEXTURE_2D, 0);
	}
	
	public String getFilename() {
		return filename;
	}
}

File: src/main/kotlin/com/morph/engine/graphics/Texture.kt
package com.morph.engine.graphics

import com.morph.engine.util.IOUtils
import org.lwjgl.BufferUtils
import org.lwjgl.opengl.GL11.*
import org.lwjgl.opengl.GL12.GL_CLAMP_TO_EDGE
import org.lwjgl.opengl.GL13.GL_TEXTURE0
import org.lwjgl.opengl.GL13.glActiveTexture
import org.lwjgl.stb.STBTTBakedChar
import org.lwjgl.stb.STBTruetype
import java.io.IOException
import java.nio.ByteBuffer
import java.util.*
import javax.imageio.ImageIO

class Texture {
    val filename: String
    private val resource: TextureResource? // TODO: Make non-nullable, an error in texture loading should return an empty Texture

    constructor(filename: String?) {
        this.filename = filename ?: "textures/solid.png"
        val oldResource = loadedTextures[this.filename]
        if (oldResource != null) {
            this.resource = oldResource
            resource.addReference()
        } else {
            resource = loadTexture(this.filename)
            loadedTextures[this.filename] = resource!!
        }
    }

    constructor(font: String?, size: Int, bitmapWidth: Int, bitmapHeight: Int) {
        this.filename = font ?: "C:/Windows/Fonts/arial.ttf"
        val oldResource = loadedTextures[this.filename + ":" + size]
        if (oldResource != null) {
            this.resource = oldResource
            resource.addReference()
        } else {
            resource = loadFont(this.filename, size, bitmapWidth, bitmapHeight)
            loadedTextures[this.filename] = resource
        }
    }

    constructor(id: String, width: Int, height: Int, pixels: ByteBuffer) {
        this.filename = id
        val oldResource = loadedTextures[this.filename]
        if (oldResource != null) {
            this.resource = oldResource
            resource.addReference()
        } else {
            resource = loadTextureFromByteBuffer(width, height, pixels)
            loadedTextures[this.filename] = resource
        }
    }

    private fun loadTextureFromByteBuffer(width: Int, height: Int, buffer: ByteBuffer): TextureResource {
        val id = glGenTextures()
        buffer.flip()

        glBindTexture(GL_TEXTURE_2D, id)

        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE)
        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE)

        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR)
        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR)

        glPixelStorei(GL_UNPACK_ALIGNMENT, 1)

        glTexImage2D(GL_TEXTURE_2D, 0, GL_ALPHA, width, height, 0, GL_ALPHA, GL_UNSIGNED_BYTE, buffer)

        return TextureResource(id)
    }

    private fun loadTexture(filename: String): TextureResource? {
        try {
            val image = ImageIO.read(Texture::class.java.classLoader.getResourceAsStream(filename))
            val pixels = image.getRGB(0, 0, image.width, image.height, null, 0, image.width)

            val buffer = BufferUtils.createByteBuffer(image.width * image.height * 4)

            val hasAlpha = image.colorModel.hasAlpha()

            for (y in 0 until image.height) {
                for (x in 0 until image.width) {
                    val color = pixels[x + y * image.width]
                    buffer.put((color shr 16 and 0xff).toByte())
                    buffer.put((color shr 8 and 0xff).toByte())
                    buffer.put((color and 0xff).toByte())

                    if (hasAlpha)
                        buffer.put((color shr 24 and 0xff).toByte())
                    else
                        buffer.put(0xff.toByte())
                }
            }

            buffer.flip()

            val id = glGenTextures()

            glBindTexture(GL_TEXTURE_2D, id)

            glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_REPEAT)
            glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_REPEAT)

            glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR)
            glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR)

            glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA8, image.width, image.height, 0, GL_RGBA, GL_UNSIGNED_BYTE, buffer)

            return TextureResource(id)
        } catch (e: IOException) {
            e.printStackTrace()
        }

        return null
    }

    private fun loadFont(font: String, size: Int, bitmapWidth: Int, bitmapHeight: Int): TextureResource {
        val texture = glGenTextures()
        val chars = STBTTBakedChar.malloc(96)

        try {
            val fontBuffer = IOUtils.getFileAsByteBuffer(font, 160 * 1024)
            val bitmap = BufferUtils.createByteBuffer(bitmapWidth * bitmapHeight)

            STBTruetype.stbtt_BakeFontBitmap(fontBuffer, size.toFloat(), bitmap, bitmapWidth, bitmapHeight, 32, chars)

            glBindTexture(GL_TEXTURE_2D, texture)
            glTexImage2D(GL_TEXTURE_2D, 0, GL_ALPHA, bitmapWidth, bitmapHeight, 0, GL_ALPHA, GL_UNSIGNED_BYTE, bitmap)
            glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR)
            glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR)
        } catch (e: IOException) {
            throw RuntimeException(e)
        }

        return TextureResource(texture)
    }

    fun destroy() {
        resource!!.removeReference()
    }

    fun bind() {
        glActiveTexture(GL_TEXTURE0)
        glBindTexture(GL_TEXTURE_2D, resource!!.id)
    }

    fun bind(i: Int) {
        glActiveTexture(GL_TEXTURE0 + i)
        glBindTexture(GL_TEXTURE_2D, resource!!.id)
    }

    fun unbind() {
        glBindTexture(GL_TEXTURE_2D, 0)
    }

    companion object {
        private val loadedTextures = HashMap<String, TextureResource>()
    }
}
